onnxruntime/core/optimizer/transpose_optimizer/transpose_optimizer.cc

```diff
@@ -1418,74 +1418,119 @@ static bool HandleTile(HandlerArgs& args) {
-
-      if (transpose_inp_node != nullptr && transpose_inp_consumers->comprehensive) {
-        // Will move output to parent. First replace parent references with name of 2nd transpose output.
-        args.node.SetInput(0, "");
-        ReplaceValueReferences(transpose_inp_consumers->nodes, transpose_input, node_output);
-        const std::vector<std::string_view>& transpose_inp_outputs = transpose_inp_node->Outputs();
-
-        // Find index of output from parent node
-        size_t i;
-        for (i = 0; i < transpose_inp_outputs.size(); ++i) {
-          if (transpose_inp_outputs[i] == transpose_input) break;
-        }
-
-        // Move 2nd transpose output (possible graph output) over top of it.
-        args.ctx.graph.MoveOutput(args.node, 0, *transpose_inp_node, i);
-      } else {
-        // Worst-case scenario: Both parent output and 2nd transpose output cannot be removed (both graph outputs)
-        // despite computing the same value. Use an Identity op instead.
-        std::vector<std::string_view> single_empty_input{""};
-        auto identity_ptr = args.ctx.graph.AddNode("Identity", single_empty_input, /*num_outputs*/ 1);
-        api::NodeRef& identity = *identity_ptr;
-        args.ctx.graph.MoveOutput(args.node, 0, identity, 0);
-        identity.SetInput(0, transpose_input);
-      }
-    }
-    // In any case, the 2nd transpose can be removed.
-    args.ctx.graph.RemoveNode(args.node);
+    RemoveCancelingTransposeNodes(args);
   } else {
     // Case 2: Permutations don't cancel. Compose permutations.
     std::vector<int64_t> new_perm = ComposePerm(args.perm, *node_perm);
     args.node.SetAttributeInts("perm", new_perm);
-    args.node.SetInput(0, transpose_input);
-  }
+    args.node.SetInput(0, args.transpose.Inputs()[0]);
 
-  // 2nd transpose no longer references 1st. Remove first if possible.
-  if (!args.ctx.graph.HasValueConsumers(args.transpose.Outputs()[0])) {
-    args.ctx.graph.RemoveNode(args.transpose);
+    // 2nd transpose no longer references 1st. Remove first if possible.
+    if (!args.ctx.graph.HasValueConsumers(args.transpose.Outputs()[0])) {
+      args.ctx.graph.RemoveNode(args.transpose);
+    }
   }
 
   return true;
 }
 
 constexpr HandlerInfo transpose_handler = {&FirstInput, &HandleTranspose, /*transposes_outputs*/ false};
 
+static bool HandleReshape(HandlerArgs& args) {
+  // We check for a very specific case where Tranpose is replaced by Reshape 
+  // for performance. For example Transpose(input {1, 1, 1, X}, perm{0, 3, 2, 1}) can be replaced by Reshape
+  // Reshape(input{1, 1, 1, X}, shape{1, X, 1, 1})
+  // During tranpose optimization we need to detect such reshape nodes so that we can remove them if possible.
+
+  // Get shape input of reshape node
+  auto shape_data = args.ctx.graph.GetConstant(args.node.Inputs()[1]);
+  if (shape_data == nullptr || shape_data->Data().size() == 0) {
+    return false;
+  }
+
```

Comment:
Do you need to check that there's only one dim that is not == 1?

Reply:
Consider X->Transpose->Y->Reshape->Z, we check shape(X) == shape(Z), that is why I did not add another check for validating only 1 dim is not 1.

Reply:
Doesn't that assume the Reshape is always undo-ing the Transpose? 

If more than one dim has a value of 1, the Reshape is not equivalent to a Transpose. Not sure if that occurs in a model, but it seems risky for that not to be explicitly validated.

Reply:
Right, we remove the 2 nodes only when they cancel each other out. In all other cases we keep them.